File: frontend/demo/component/combobox/combo-box-custom-entry-2.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/combo-box';
import type { ComboBoxCustomValueSetEvent } from '@vaadin/combo-box';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('combo-box-custom-entry-2')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private items = ['Chrome', 'Edge', 'Firefox', 'Safari'];

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-combo-box
        allow-custom-value
        label="Browser"
        helper-text="Select or type a browser"
        .items="${this.items}"
        @custom-value-set="${(event: ComboBoxCustomValueSetEvent) => {
          this.items = [...this.items, event.detail];
        }}"
      ></vaadin-combo-box>
      <!-- end::snippet[] -->
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/combobox/ComboBoxCustomEntry2.java
package com.vaadin.demo.component.combobox;

import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.router.Route;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("combo-box-custom-entry-2")
public class ComboBoxCustomEntry2 extends Div {

  private List<String> items = new ArrayList<>(Arrays.asList("Chrome", "Edge", "Firefox", "Safari"));

  public ComboBoxCustomEntry2() {
    // tag::snippet[]
    ComboBox<String> comboBox = new ComboBox<>("Browser");
    comboBox.setAllowCustomValue(true);
    comboBox.addCustomValueSetListener(e -> {
      String customValue = e.getDetail();
      items.add(customValue);
      comboBox.setItems(items);
      comboBox.setValue(customValue);
    });
    add(comboBox);
    // end::snippet[]
    comboBox.setItems(items);
    comboBox.setHelperText("Select or type a browser");
  }

  public static class Exporter extends DemoExporter<ComboBoxCustomEntry2> { // hidden-source-line
  } // hidden-source-line
}
